src/transport/middleware.rs

```diff
@@ -129,7 +122,29 @@ where
             );
         }
 
-        self.transport.send(req).await
+        Ok(req)
+    }
+}
+
+#[async_trait::async_trait]
+impl<T> Transport for PubNubMiddleware<T>
+where
+    T: Transport,
+{
+    async fn send(&self, req: TransportRequest) -> Result<TransportResponse, PubNubError> {
+        self.prepare_request(req)
+            .map(|req| self.transport.send(req))?
```

Comment:
`map` maps the request into `Pin<Box<dyn Future>>` that must be handled.
`and_then` doing the same thing but adds a layer of result. 

For async is easier to use map because you don't have to wrap it into `Ok()`. 
For blocking `and_then` work well with the return type so returns proper one. 
It can be archived by using
`.map(|req| self.transport.send(req))?`
 instead of 
`.and_then(|req| self.transport.send(req))`

Reply:
explained at dm